src/accelerate/local_sgd.py

```diff
@@ -0,0 +1,104 @@
+    back to at least:
+
+    Zhang, J., De Sa, C., Mitliagkas, I., & Ré, C. (2016). [Parallel SGD: When does averaging help?. arXiv preprint
+    arXiv:1606.07365.](https://arxiv.org/abs/1606.07365)
+
+    We credit the term Local SGD to the following paper (but there might be earlier references we are not aware of).
+
+    Stich, Sebastian Urban. ["Local SGD Converges Fast and Communicates Little." ICLR 2019-International Conference on
+    Learning Representations. No. CONF. 2019.](https://arxiv.org/abs/1805.09767)
+
+    """
+
+    def __enter__(self):
+        if self.enabled:
+            self.model_sync_obj = self.model.no_sync()
+            self.model_sync_obj.__enter__()
+
+        return self
+
+    def __exit__(self, type, value, tb):
+        if self.enabled:
+            # Average all models on exit
+            self._sync_and_avg_model_params()
+            self.model_sync_obj.__exit__(type, value, tb)
+
+    def __init__(self, accelerator: Accelerator, model: torch.nn.Module, local_sgd_steps: int, enabled: bool = True):
+        """
+        Constructor.
+
+        Args:
+            model (`torch.nn.Module):
+                The model whose parameters we need to average.
+            accelerator (`Accelerator`):
+                Accelerator object.
+            local_sgd_steps (`int`):
+                A number of local SGD steps (before model parameters are synchronized).
+            enabled (`bool):
+                Local SGD is disabled if this parameter set to `False`.
+        """
+        if accelerator.distributed_type not in [
+            DistributedType.NO,
+            DistributedType.MULTI_CPU,
+            DistributedType.MULTI_GPU,
+        ]:
+            raise NotImplementedError("LocalSGD is supported only for CPUs and GPUs (no DeepSpeed or MegatronLM)")
+        self.enabled = enabled and accelerator.distributed_type != DistributedType.NO
+        self.num_steps = 0
+        if self.enabled:
+            self.accelerator = accelerator
+            self.model = model
+            self.local_sgd_steps = local_sgd_steps
+
+    def step(self):
+        """
+        This function makes a "step" and synchronizes model parameters if necessary.
+        """
+        self.num_steps += 1
+        if not self.enabled:
+            return
+
+        if self.num_steps % self.local_sgd_steps == 0:
+            self._sync_and_avg_model_params()
+
+    def _sync_and_avg_model_params(self):
+        """
+        Synchronize + Average model parameters across all GPUs
+        """
+        import torch.distributed as dist
+
+        self.accelerator.wait_for_everyone()
+        with self.accelerator.autocast():
+            qty = float(dist.get_world_size())
+            for param in self.model.parameters():
+                dist.all_reduce(param.data, op=torch.distributed.ReduceOp.SUM)
+                param.data /= qty
```

Comment:
All of this should be able to be replaced with:

```suggestion
            for param in self.model.parameters():
                self.accelerator.reduce(param.data)
```

Reply:
@searchivarius the only thing I can think of off the top of my head is:

Do you still experience this if you do:
```python
            qty = int(dist.get_world_size())
            for param in self.model.parameters():
                dist.all_reduce(param.data, op=torch.distributed.ReduceOp.SUM)
                param.data /= qty
```

(Notice `qty` is `int` instead of `float`)

Reply:
As otherwise, the two should be the exact same.

Reply:
@searchivarius I see what's up, you need to pass `reduction="mean"` during `accelerator.reduce`

Reply:
Hi @muellerzr this still doesn't work this way. However, after looking more carefully at the code, I realized that the operation is done on the cloned tensor. So the suggested reduction operation does nothing. The reduced tensor needs to be reassigned back. Fortunately, this seems to be finally working:
```
-                self.accelerator.reduce(param.data, reduction="mean")
+                param.data = self.accelerator.reduce(param.data, reduction="mean")
```

Reply:
PS: also yet another piece of evidence of how Python is really bad language that doesn't do useful type checks for you.